avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDatumReaderWriterUtil.java

```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      SchemaPair that = (SchemaPair) o;
+      return writerSchema == that.writerSchema && readerSchema == that.readerSchema;
+    }
+
+    @Override
+    public int hashCode() {
+      return hashCode;
+    }
+  }
+
+  private static final ReentrantReadWriteLock reentrantReadWriteLock = new ReentrantReadWriteLock();
+
+  //TODO :  LRU cache
+  private static final Map<SchemaPair, FastGenericDatumReader> fastGenericDatumReaderCache = new ConcurrentHashMap<>();
+  private static final Map<Schema, FastGenericDatumWriter> fastGenericDatumWriterCache = new WeakIdentityHashMap();
+
+  private static final Map<SchemaPair, FastSpecificDatumReader> fastSpecificDatumReaderCache = new ConcurrentHashMap<>();
+  private static final Map<Schema, FastSpecificDatumWriter> fastSpecificDatumWriterCache = new WeakIdentityHashMap();
+
+  private FastDatumReaderWriterUtil() {
+  }
+
+  public static FastGenericDatumReader getFastGenericDatumReader(Schema schema) {
+    return getFastGenericDatumReader(schema, schema);
+  }
+
+  public static FastGenericDatumReader getFastGenericDatumReader(Schema writerSchema, Schema readerSchema) {
+    SchemaPair schemaPair = new SchemaPair(writerSchema, readerSchema);
+    return fastGenericDatumReaderCache.computeIfAbsent(schemaPair, key -> new FastGenericDatumReader<>(writerSchema, readerSchema));
+  }
+
+  public static FastGenericDatumWriter getFastGenericDatumWriter(Schema writerSchema) {
+    FastGenericDatumWriter fastDatumWriter = null;
+
+    // lookup cache and read lock
+    reentrantReadWriteLock.readLock().lock();
+    try {
+      fastDatumWriter = fastGenericDatumWriterCache.get(writerSchema);
+    } finally {
+      reentrantReadWriteLock.readLock().unlock();
+    }
+    // update cache and write lock
+    if (fastDatumWriter == null) {
+      try {
+        reentrantReadWriteLock.writeLock().lock();
+        fastDatumWriter = new FastGenericDatumWriter<>(writerSchema);
+        fastGenericDatumWriterCache.put(writerSchema, fastDatumWriter);
+      } finally {
+        reentrantReadWriteLock.writeLock().unlock();
+      }
+    }
+    return fastDatumWriter;
+  }
+
+
+  public static FastSpecificDatumReader getFastSpecificDatumReader(Schema schema) {
+    return getFastSpecificDatumReader(schema, schema);
+  }
+
+  public static FastSpecificDatumReader getFastSpecificDatumReader(Schema writerSchema, Schema readerSchema) {
+    SchemaPair schemaPair = new SchemaPair(writerSchema, readerSchema);
+    return fastSpecificDatumReaderCache.computeIfAbsent(schemaPair, key -> new FastSpecificDatumReader<>(writerSchema, readerSchema));
+  }
+
+  public static FastSpecificDatumWriter getFastSpecificDatumWriter(Schema writerSchema) {
```

Comment:
This function is almost same as `getFastGenericDatumWriter`, so we can create a utility function to wrap up the common logic, and the common function could take a `supplier` callback to create the actual DatumWriter.

Reply:
Yes . The functions are identical . We use different cache for generic and specific types though .

Reply:
That is right, but it is still possible to pass the suppliers to define how to retrieve from the cache/put into the cache.
The goal here is trying to minimize  duplicate  codes.
The common function could be:
DatumWriter<T> getFastDatumWriter(Schema schema, Supplier<DatumWriter> readFromCache, Consumer<DatumWriter> putIntoCache);
With this way, you need to do some static casting to make it work.

BTW, why did you remove the generic type in the signature of all those functions?

Reply:
Thanks for pointing this out @gaojieliu  . Have used generic types .  
Totally agree with the points about having a common getFastDatumWriter()  . Will take that up as an improvement .